workflow/src/main/java/bio/terra/tanagra/workflow/CopyBigQueryTable.java

```diff
@@ -0,0 +1,144 @@
+package bio.terra.tanagra.workflow;
+
+import bio.terra.tanagra.proto.underlay.Column;
+import bio.terra.tanagra.proto.underlay.Dataset.BigQueryDataset;
+import bio.terra.tanagra.proto.underlay.Table;
+import com.google.api.services.bigquery.model.TableReference;
+import com.google.api.services.bigquery.model.TableRow;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.sql.DataSource;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
+import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.TypedRead.Method;
+import org.apache.beam.sdk.io.jdbc.JdbcIO;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+
+/**
+ * Copies the contents of a BigQuery table to a SQL table.
+ *
+ * <p>The SQL table must already exist.
+ */
+@AutoValue
```

Comment:
Interesting that you're able to Autovalue in combination with PTransform; does the annotation basically just get applied to the generated class as well?

Relatedly, for my own benefit: do you know off-hand or have a guess about whether this can work with Hibernate entities?

Reply:
I'm not sure which annotation you mean. For a PTransform, the `expand` is overriding the base PTransform function, unlike some of the DoFn or other function's in joins/JdbcIO calls that are based on annotations. The AutoValue generated class won't create it's own `expand`.
I mostly went with making it an autovalue for configurations and some casual reading of https://beam.apache.org/contribute/ptransform-style-guide/#fluent-builders-for-configuration , which is maybe a little more for library functions than what we're strictly doing here.

I'm not sure what you're asking with Hibernate entities.
If this is a question about Hibernate in Beam, there's some implicit serialization going on in all of these pipelines as data gets distributed to different VMs. Having not worked with Hibernate, I'd think it's POJO classes would do fine with serialization. I don't know how the Hibernate<->JDBC connection would as a part of pipeline transformations.

If the question was about inheriting annotations more generally (e.g. AutoValue classes for Hibernate `@Entity` clases), I think it behavior changes based on if the annotation itself is marked as[ java.lang.annotation.Inherited](https://docs.oracle.com/javase/7/docs/api/java/lang/annotation/Inherited.html). But AutoValue has some support for moving annotations along to the generated implementation classes https://github.com/google/auto/blob/master/value/userguide/howto.md#-copy-annotations-from-a-classmethod-to-the-implemented-classmethodfield. So, sounds feasible, but I haven't mucked with that before myself.

Reply:
Sorry, rereading - my annotation question didn't make sense, my brain jumped ahead to how I could use this in combination with the `@Entity` annotation.

The latter is what I was asking about - thanks for the link.